server/src/main/java/org/elasticsearch/cluster/coordination/StableMasterHealthIndicatorService.java

```diff
@@ -0,0 +1,449 @@
+        String summary = String.format(
+            Locale.ROOT,
+            "The master has changed %d times in the last %s",
+            masterChanges,
+            localMasterHistory.getMaxHistoryAge()
+        );
+        HealthIndicatorDetails details = getSimpleDetails(explain, localMasterHistory);
+        Collection<HealthIndicatorImpact> impacts = getUnstableMasterImpacts();
+        List<UserAction> userActions = getContactSupportUserActions(explain);
+        return createIndicator(stableMasterStatus, summary, explain ? details : HealthIndicatorDetails.EMPTY, impacts, userActions);
+    }
+
+    /**
+     * This returns HealthIndicatorDetails.EMPTY if explain is false, otherwise a HealthIndicatorDetails object containing only a
+     * "current_master" object and a "recent_masters" array. The "current_master" object will have "node_id" and "name" fields for the
+     * master node. Both will be null if the last-seen master was null. The "recent_masters" array will contain "recent_master" objects.
+     * Each "recent_master" object will have "node_id" and "name" fields for the master node. These fields will never be null because
+     * null masters are not written to this array.
+     * @param explain If true, the HealthIndicatorDetails will contain "current_master" and "recent_masters". Otherwise it will be empty.
+     * @param localMasterHistory The MasterHistory object to pull current and recent master info from
+     * @return An empty HealthIndicatorDetails if explain is false, otherwise a HealthIndicatorDetails containing only "current_master"
+     * and "recent_masters"
+     */
+    private HealthIndicatorDetails getSimpleDetails(boolean explain, MasterHistory localMasterHistory) {
+        return explain ? (builder, params) -> {
+            builder.startObject();
+            DiscoveryNode masterNode = localMasterHistory.getMostRecentMaster();
+            builder.object(DETAILS_CURRENT_MASTER, xContentBuilder -> {
+                if (masterNode != null) {
+                    builder.field("node_id", masterNode.getId());
+                    builder.field("name", masterNode.getName());
+                } else {
+                    builder.nullField("node_id");
+                    builder.nullField("name");
+                }
+            });
+            List<DiscoveryNode> recentMasters = localMasterHistory.getNodes();
+            builder.array(DETAILS_RECENT_MASTERS, arrayXContentBuilder -> {
+                for (DiscoveryNode recentMaster : recentMasters) {
+                    if (recentMaster != null) {
+                        builder.startObject();
+                        builder.field("node_id", recentMaster.getId());
+                        builder.field("name", recentMaster.getName());
+                        builder.endObject();
+                    }
+                }
+            });
+            return builder.endObject();
+        } : HealthIndicatorDetails.EMPTY;
```

Comment:
I think we should limit ternary operators to single-line statements, since it's difficult to parse as a human when it spans tens of lines.

Perhaps we could do something like:

```java
if (explain == false) {
    return HealthIndicatorDetails.EMPTY;
} else {
    return <this big thing>;
}
```